nntrainer/src/neuralnet.cpp

```diff
@@ -213,252 +150,119 @@ int NeuralNetwork::loadFromConfig() {
-      NN_INI_RETURN_STATUS();
-
-      pooling2d_layer->setPoolingType(
-        (nntrainer::Pooling2DLayer::PoolingType)parseType(
-          iniparser_getstring(ini, (layer_name + ":pooling").c_str(),
-                              "average"),
-          TOKEN_POOLING));
-
-      addLayer(pooling2d_layer);
-    } break;
-
-    case LAYER_FLATTEN: {
-      std::shared_ptr<FlattenLayer> flatten_layer =
-        std::make_shared<FlattenLayer>();
-
-      status = flatten_layer->setName(layer_name);
-      NN_INI_RETURN_STATUS();
+  return status;
+}
 
-      addLayer(flatten_layer);
-    } break;
+int NeuralNetwork::loadFromConfig() {
+  int status = ML_ERROR_NONE;
+  std::string ini_file = config;
+  int num_ini_sec = 0;
+  dictionary *ini;
 
-    case LAYER_FC: {
-      WeightDecayParam weight_decay;
-      std::shared_ptr<FullyConnectedLayer> fc_layer =
-        std::make_shared<FullyConnectedLayer>();
+  if (ini_file.empty()) {
+    ml_loge("Error: Configuration File is not defined");
+    return ML_ERROR_INVALID_PARAMETER;
+  }
 
-      status = fc_layer->setName(layer_name);
-      NN_INI_RETURN_STATUS();
+  /** Parse ini file */
+  ini = iniparser_load(ini_file.c_str());
+  if (ini == NULL) {
+    ml_loge("Error: cannot parse file: %s\n", ini_file.c_str());
+    return ML_ERROR_INVALID_PARAMETER;
+  }
 
-      std::string input_shape_str = iniparser_getstring(
-        ini, (layer_name + ":Input_Shape").c_str(), unknown);
-
-      if (input_shape_str.compare("Unknown") != 0) {
-        TensorDim d;
-        d.setTensorDim(input_shape_str);
-        fc_layer->setInputDimension(d);
-      } else if (section_names_iter == section_names.begin()) {
-        ml_loge("Error: %s layer input shape not specified.",
-                layer_name.c_str());
-        status = ML_ERROR_INVALID_PARAMETER;
-        NN_INI_RETURN_STATUS();
-      }
+  /** Get number of sections in the file */
+  num_ini_sec = iniparser_getnsec(ini);
+  if (num_ini_sec < 0) {
+    ml_loge("Error: invalid number of sections.");
+    return ML_ERROR_INVALID_PARAMETER;
+  }
 
-      fc_layer->setUnit(static_cast<unsigned int>(
-        iniparser_getint(ini, (layer_name + ":Unit").c_str(), 0)));
+  /** Get all the section names */
+  for (int idx = 0; idx < num_ini_sec; ++idx) {
+    const char *sec_name = iniparser_getsecname(ini, idx);
 
-      fc_layer->setBiasZero(b_zero);
-      fc_layer->setWeightInit((WeightIniType)parseType(
-        iniparser_getstring(ini, (layer_name + ":WeightIni").c_str(),
-                            "xavier_uniform"),
-        TOKEN_WEIGHTINI));
+    if (!sec_name) {
+      ml_loge("Error: Unable to retrieve section names from ini.");
+      return ML_ERROR_INVALID_PARAMETER;
+    }
 
-      status = parseWeightDecay(ini, layer_name, weight_decay);
-      NN_INI_RETURN_STATUS();
+    if (strncasecmp("network", sec_name, 7) == 0) {
+      status = loadNetworkConfig((void *)ini);
```

Comment:
with this implementation, we state that a valid ini file can be without `netwoork` section. Is that intended?

Reply:
The code undergoes validation starting from `:298` it would fail if there is no network section.

It is moved to segregate validation phase for better readability (as this function is not critical w.r.t performance)